lib/auth/join_iam.go

```diff
@@ -49,16 +53,103 @@ const (
 	// ever have a need to allow a newer API version.
 	expectedStsIdentityRequestBody = "Action=GetCallerIdentity&Version=2011-06-15"
 
-	// Only allowing the global sts endpoint here, Teleport nodes will only send
-	// requests for this endpoint. If we want to start using regional endpoints
-	// we can update this check before updating the nodes.
-	stsHost = "sts.amazonaws.com"
+	// Used to check if we were unable to resolve the regional STS endpoint.
+	globalSTSEndpoint = "https://sts.amazonaws.com"
 
 	// AWS SignedHeaders will always be lowercase
 	// https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html#sigv4-auth-header-overview
 	challengeHeaderKey = "x-teleport-challenge"
 )
 
+// validateSTSHost returns an error if the given stsHost is not a valid regional
+// endpoint for the AWS STS service, or nil if it is valid.
+//
+// This is a security-critical check: we are allowing the client to tell us
+// which URL we should use to validate their identity. If the client could pass
+// off an attacker-controlled URL as the STS endpoint, the entire security
+// mechanism of the IAM join method would be compromised.
+//
+// The simplest approach would be to make sure that the URL matches
+// "sts.(<region>.)?amazonaws.com(.cn)?" but this fails in a couple of ways:
+// - it seems that, at least in the past, Amazon was selling subdomains of
+//   "amazonaws.com", so we don't know who might control that endpoint.
+// - Second, it does not match endpoints in the AWS ISO partitions such as
+//   "sts.us-iso-east-1.c2s.ic.gov". We don't know if AWS will add more STS
+//   endpoints like this with different URL patterns.
+//
+// Probably the most secure approach would be to check the given STS URL against
+// a static list of known STS endpoints. This would be hard to maintain over
+// time as AWS adds new regions and partitions.
+//
+// Another option would be to query AWS at runtime to get the list of valid STS
+// endpoints. I couldn't find a suitable API to support this.
+//
+// The approach I've chosen here basically offloads the task of maintaining the
+// static list of STS endpoints to the AWS SDK. Since I can't get at the list
```

Comment:
Any concerns about a supply chain attack where someone tricks us into using a modified  version of the AWS SDK and thus using an untrusted endpoint?

I think I'd be okay with it requiring a patch release to support new regions and partitions if it's easier to secure with a hard-coded list.

I believe what you have implemented here to be secure. But there's a fair amount of logic and a lot of special cases to handle, which makes me nervous that we may slip up in the future.

Reply:
The AWS SDK is compiled into the `teleport` binary, if someone can trick us into using a modified SDK I think there would be much easier and worse ways to pwn us.

But you're right, the validation ended up being more finicky than I would like, and a static list wouldn't be that difficult to maintain, I'll switch to that

Reply:
Done in the latest